crates/pathfinder/src/merkle_tree.rs

```diff
@@ -0,0 +1,1089 @@
+    ///
+    /// This implies decrementing the root's reference count. The root will
+    /// only get deleted if the reference count reaches zero. This will in turn
+    /// delete all internal nodes and leaves which no longer have a root to connect to.
+    ///
+    /// This allows for multiple instances of the same tree state to be committed,
+    /// without deleting all of them in a single call.
+    pub fn delete(self) -> anyhow::Result<()> {
+        match self.root.borrow().hash() {
+            Some(hash) if hash != ZERO_HASH => self
+                .storage
+                .decrement_ref_count(hash)
+                .context("Failed to delete tree root"),
+            _ => Ok(()),
+        }
+    }
+
+    /// Loads an existing tree or creates a new one if it does not yet exist.
+    ///
+    /// Use the [ZERO_HASH] as root if the tree does not yet exist, will otherwise
+    /// error if the given hash does not exist.
+    ///
+    /// The transaction is used for all storage interactions. The transaction
+    /// should therefore be committed after all tree mutations are completed.
+    ///
+    /// Uses an [RcNodeStorage] as backing storage.
+    ///
+    /// ### Warning
+    ///
+    /// None of the [RcNodeStorage] functions rollback on failure. This means that if any error
+    /// is encountered, the transaction should be rolled back to prevent database corruption.
+    pub fn load(
+        table: String,
+        transaction: &'a Transaction,
+        root: StarkHash,
+    ) -> anyhow::Result<Self> {
+        let storage = RcNodeStorage::open(table, transaction)?;
+
+        // Create a tree with root node unresolved, so we can use the resolve function.
+        // Bit clumsy, but oh well.
+        let root_node = Rc::new(RefCell::new(Node::Unresolved(root)));
+        let mut tree = Self {
+            storage,
+            root: root_node,
+        };
+
+        if root != ZERO_HASH {
+            // Resolve non-zero root node to check that it does exist.
+            let root_node = tree
+                .resolve(root, 0)
+                .context("Failed to resolve root node")?;
+            tree.root = Rc::new(RefCell::new(root_node));
+        }
+
+        Ok(tree)
+    }
+
+    /// Persists all changes to storage and returns the new root hash.
+    ///
+    /// Note that the root is reference counted in storage. Committing the
+    /// same tree again will therefore increment the count again.
+    pub fn commit(self) -> anyhow::Result<StarkHash> {
+        // Go through tree, collect dirty nodes, calculate their hashes and
+        // persist them. Take care to increment ref counts of child nodes. So in order
+        // to do this correctly, will have to start back-to-front.
+        self.commit_subtree(&mut *self.root.borrow_mut())?;
+        // unwrap is safe as `commit_subtree` will set the hash.
+        let root = self.root.borrow().hash().unwrap();
+        self.storage.increment_ref_count(root)?;
```

Comment:
On its own it is not needed, but its required because of how I implemented the `delete` tree function. In the delete function I decrement the root reference count. I do this instead of just deleting the root node outright because what could easily happen is:

1. I receive a state update at time `t=N`.
    1.  new state root = `H`
2. More state updates, time moves on....
3. I receive another update at `t=N'`. I coincidentally end up with an identical state.
     2. new state root also = `H`

At some point I now want to delete older state. I can now call delete for state root `H` to clean up the state from `t=N`. The root `H` will still remain (since it is still referenced by the update in (3)).

Reply:
So one can think of it as external references to the root, incrementing the root nodes reference count.